design/storage-class-device-set.md

```diff
@@ -0,0 +1,478 @@
+that they can be attached to cluster nodes.
+
+A `StorageClassDeviceSet` will have the following fields associated with it:
+
+* **name**: A name for the set. **[required]**
+* **count**: The number of devices in the set. **[required]**
+* **resources**: The CPU and RAM requests/limits for the devices. Default is no
+  resource requests.
+* **placement**: The placement criteria for the devices. Default is no
+  placement criteria.
+* **config**: Granular device configuration. This is a generic
+  `map[string]string` to allow for provider-specific configuration.
+* **volumeClaimTemplates**: A list of PVC templates to use for provisioning the
+  underlying storage devices.
+
+An entry in `volumeClaimTemplates` must specify the following fields:
+  * **resources.requests.storage**: The desired capacity for the underlying
+  storage devices.
+  * **storageClassName**: The StorageClass to provision PVCs from.
+
+## Example Workflow: rook-ceph OSDs
+
+The CephCluster CRD could be extended to include a new field:
+`spec.storage.StorageClassDeviceSets`, which would be a list of one or more
+`StorageClassDeviceSets`. If elements exist in this list, the CephCluster
+controller would then create enough PVCs to match each `StorageClassDeviceSet`'s
+`Count` field, attach them to individual OsdPrepare Jobs, then attach them to
+OSD Pods once the Jobs are completed. For the initial implementation, only one
+entry in `volumeClaimTemplates` would be supported, if only to tighten the scope
+for an MVP.
+
+The PVCs would be provisioned against a configured or default StorageClass. It
+is recommended that the admin setup a StorageClass with `volumeBindingMode:
+WaitForFirstConsumer` set.
+
+If the admin wishes to control device placement, it will be up to them to make
+sure the desired nodes are labeled properly to ensure the Kubernetes scheduler
+will distribute the OSD Pods based on Placement criteria.
+
+In keeping with current Rook-Ceph patterns, the **resources** and **placement**
+for the OSDs specified in the `StorageClassDeviceSet` would override any
+cluster-wide configurations for OSDs. Additionally, the `StorageClassDeviceSet`
+configuration would take precedence over any other conflicting configurations,
+such as if `useAllDevices` is configured for the cluster.
+
+### OSD Deployment Behavior
+
+While the current strategy of deploying OSD Pods as individual Kubernetes
+Deployments, some changes to the deployment logic would need to change. The
+workflow would look something like this:
+
+1. Get all matching OSD PVCs
+1. Create any missing OSD PVCs
+1. Get all matching OSD Deployments
+1. Check that all OSD Deployments are using valid OSD PVCs
+    * If not, probably remove the OSD Deployment?
+    * Remove any PVCs used by OSD Deployments from the list of PVCs to be
+    worked on
+1. Run an OsdPrepare Job on all unused and uninitialized PVCs
+    * This may be either one Job per PVC or one Job per batch of PVCs. One per
```

Comment:
Yes, if we batch and one prepare fails then the remaining ones won't be prepared so 1 prepare per PVC.

Reply:
@jarrpa Can you change this to be clear about one job per PVC?

Reply:
Sure,